src/crystal/main.cr

```diff
@@ -32,22 +36,39 @@ module Crystal
   # same can be accomplished with `at_exit`. But in some cases
   # redefinition of C's main is needed.
   def self.main(&block)
-    GC.init
+    # initialize Crystal core foundations, responsible to initialize enough of
+    # the main user code, which includes corelib and stdlib initializations, and
+    # that will be executed in a fiber:
+    GC.init                     # memory allocator
+    Fiber.init                  # stack pool, fiber list
+    Thread.init                 # threads, main thread
+    Crystal::EventLoop.init     # I/O, sleep timers, ...
+    Crystal::Scheduler.init     # fiber schedulers (threads)
 
-    status =
+    status = 0
+
+    # start the main fiber, a scheduler thread will pick it up later:
+    spawn(name: "main_user_code") do
       begin
-        yield
-        0
+        block.call
       rescue ex
-        1
+        AtExitHandlers.exception = ex
+        status = 1
       end
 
-    AtExitHandlers.exception = ex if ex
+      status = AtExitHandlers.run(status)
+      STDOUT.flush
+      STDERR.flush
```

Comment:
Oh, good catch.

Reply:
Oh, I remember why I left 'em here: once we break out of the main user code fiber, scheduler threads will stop... but STDOUT and STDERR are buffered & nonblocking and thus depend on the schedulers running. Maybe it should be `STDOUT.sync = true; STDERR.sync = true` instead, this will flush out the buffers and allow further writes (blocking calls, but we don't care at this point)?

Reply:
Yep, setting `sync` seems like the right thing here.